### ISSUE TITLE
Consider creating `interface WorkflowView<RenderingT: Screen>`

### ISSUE THREAD

**Issue by rjrjr** [square/workflow-kotlin#606]
We _could_ create a facade interface that just wrapped a plain old View. I don't think it's worth the effort. Feature developers basically never bump into this stuff, thanks to `WorkflowViewStub` and such. That will be even more true when the UI overhaul is done, I think.

But just to scratch the nerdsnipe itch, here's a sketch of what that could be.

```kotlin
fun <RenderingT> ViewRegistry.buildView(...): WorkflowView<RenderingT>

interface WorkflowView<RenderingT> {
  fun asView(): View
  
  fun start()

  val rendering: RenderingT
  val environment: ViewEnvironment

  fun update(
    rendering: RenderingT,
    environment: ViewEnvironment
  )
  // ...
}

fun <RenderingT> View.asWorkflowView(): WorkflowView<T>? {
  // Looks for the tag and confirms the type is right? 
}
```

Still thinking this would be a distraction, but I'm wavering. Dammit Steve, are you proud that you've done this to me?

_Originally posted by @rjrjr in https://github.com/square/workflow-kotlin/pull/602#discussion_r766956989_

**Comment by rjrjr** [comment]
Oh my. This wouldn't have to be Android specific, or even classic-view specific. And it could require mapping methods that would replace most of the guts of `DecorativeViewFactory`.

```kotlin
interface WorkflowView<NativeViewT, RenderingT: Screen> {
  fun asView(): NativeViewT

  fun <RenderingT: Screen, RenderingU: Screen> map(
    transform: (RenderingT) -> RenderingU
  ): WorkflowView<NativeViewT, RenderingU>

  fun withEnvironmentMapper(
    updater: (ViewEnvironment) -> ViewEnvironment
  ):  WorkflowView<NativeViewT, RenderingU>
}
```

**Comment by rjrjr** [comment]
I am definitely _not_ going to take this on as part of #195, or hold UI 1.0 hostage to it. We'll never get rid of the experimental annotation. But this would be a really nice basis for UI 2.0.

**Comment by rjrjr** [comment]
The `NativeT` variant is impractical -- e.g., that `asView(): NativeViewT` method is really `buildView()`, which requires platform-specific arguments like `Context`.

**Comment by rjrjr** [comment]
@steve-the-edwards I couldn't resist this any longer, giving it a try this week (without the `NativeViewT` bit so far). So far it seems to clean up a _lot_ of rough edges.

**Comment by rjrjr** [comment]
Took a stab at this in #621. It feels great, except that it's hard to preserve the fix for #597.

I think that would be easier and more robust if the transforms were built into `ScreenViewFactory` itself, and if `Screen.buildView` were replaced by `Screen.getViewFactory`.  But that requires fixing #594, our broken `ViewRegistry` API.

I'm already convinced we have to do that. I guess this is the week.

**Comment by rjrjr** [comment]
Yeah, I don't think this is practical. Adding a wrapper around `View` instead of extending it directly is just too big a change. The spike was useful, though, inspired some other good ideas.

**Comment by rjrjr** [comment]
Finally landed a similar cleanup in #703.
